### PR TITLE
core: Replay database prefix

### PR DESCRIPTION
**Pull request by kim**
# Description of Changes

When replaying the log, stop if a commit is encountered which cannot be decoded. As this implies a corrupted log, truncate the log to the known-good prefix.

**Depends on #526**

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

# Expected complexity level and risk

3 -- the risk of truncating the commit log wrongly is data loss.

### CODE REVIEW

File: crates/core/src/db/relational_db.rs
Comment (kim) [inline crates/core/src/db/relational_db.rs:106]: Since this is destructive, I'm wondering if we want to add more safeguards against data loss here. And if so, which ones.

We could, for example, only reset the log if the offset falls into the last segment, and/or is within a certain distance from `open_segment_max_offset` (for some definition of "certain distance").

File: crates/core/src/db/relational_db.rs
Comment (kim) [inline crates/core/src/db/relational_db.rs:102]: Note that I intend to reduce the rightward drift here and move the replay logic elsewhere.
I left it here for this patch to preserve the side-by-side diff for review.

File: crates/core/src/db/relational_db.rs
Comment (kim) [inline crates/core/src/db/relational_db.rs:1478]: It would be nice to also test the "safeguard" condition (log corrupted in the middle), but the default segment size in 1GiB. Punting on making this configurable at this point.

File: crates/core/src/db/relational_db.rs
Comment (kim) [inline crates/core/src/db/relational_db.rs:1514]: It would be nice to also test the "safeguard" condition (log corrupted in the middle), but the default segment size in 1GiB. Punting on making this configurable at this point.

Comment (kim) [reply crates/core/src/db/relational_db.rs:1514]: Actually, that’s not true but obscured by the various test helpers. Will add a test before landing.

File: crates/core/src/db/relational_db.rs
Comment (kulakowski) [inline crates/core/src/db/relational_db.rs:109]: Maybe for later: it feels like the RelationalDB should not be the place that knows this fact.

**Review by kulakowski — APPROVED** [review]
lgtm pending Joshua's feedback.

File: crates/core/src/db/relational_db.rs
Comment (kim) [reply crates/core/src/db/relational_db.rs:109]: I agree -- I have more patches on the stack to move most of the replaying into `CommitLog`. But wanted to keep the diff small for this one.

File: crates/core/src/db/relational_db.rs
Comment (kim) [inline crates/core/src/db/relational_db.rs:127]: Turns out we should've checked that this is contiguous.

File: crates/core/src/db/relational_db.rs
Comment (kim) [reply crates/core/src/db/relational_db.rs]: I did consider this, but it turns out we need to handle more cases here:

1. The `Commit` envelope decodes successfully, but is actually invalid
2. The `Commit` envelope fails to decode
3. The file is EOF _before_ commit decoding even starts

It's not obvious how to unify this into a single error type (at least not without some major refactoring), and I'm not even sure it would be of value -- we would not gain any exhaustiveness checking.

I hope to have improved the upstream error reporting (i.e. what comes out of `RelationalDB::open`) in the recent commits, though.

Comment (kulakowski) [reply crates/core/src/db/relational_db.rs]: This line of thought ok with you Joshua?

Comment (kim) [reply crates/core/src/db/relational_db.rs]: I have formulated this differently in #554, which is a large enough diff to submit separately:

https://github.com/clockworklabs/SpacetimeDB/pull/554/files#diff-37c945f95a02769137efa2981278629b8e8aa25d24504f0c4210624b552dda93R473

This keeps enough state to keep handling of the different cases close to each other. It may also become general iteration at some point, but I’m not sure.

**Pull request merged**
